modules/fluid_properties/include/userobjects/SinglePhaseFluidProperties.h

```diff
@@ -277,7 +277,7 @@ class SinglePhaseFluidProperties : public FluidProperties
    * @return internal energy (J/kg)
    */
   virtual Real e_from_p_T(Real p, Real T) const;
-  virtual DualReal e_from_p_T(DualReal p, DualReal T) const;
+  DualReal e_from_p_T(const DualReal & p, const DualReal & T) const;
```

Comment:
Are these not virtual, because the way you compute it is always the same and you actually rely on calling the virtual versions inside the methods?  If so, we need this to be mentioned in the doco for this class.

Reply:
I did this to be consistent with the definitions in `SinglePhaseFluidProperties.h` in the `propfunc` macro for the other methods. 

I'm not sure that this is the most flexible design though. For the simple fluids like the ideal gas, it is trivial to do the derivatives, so this idea makes sense. For more complicated fluids (like the ones I use), the derivatives are a pain to code, so it might be nice to code the DualReal version instead.

Reply:
Let me demonstrate the idea behind `propfunc`:

1. let's say we need to compute `e_from_p_T` and our primary variables are `rho`, `rhou`, `rhoE`.
2. The `DualReal` for `p` has this structure: `(value of p, dp_drho, dp_drhou, dp_rhoE)`, similarly for `T`.
2. The value we need back for `e` is: `(value of e, de_drho, de_drhou, de_drhoE)`.
3. So we need to compute: `(e_from_p_T, de_dp * dp_drho + de_dT * dT_drho, de_dp * dp_drhou + de_dT * dT_drhou, de_dp * dp_drhoE + de_dT * dT_drhoE)`.

The `d{p|T}_d{rho|rhou|rhoE}` is coming from the AD system and `de_dp` and `de_dT` must be computed.

If `de_dp` and `de_dT` are difficult to compute (as you say), I am not sure that computing the `DualReal` version would be easier. 

Also, if we are solving in terms of `p` and `T`, then the `dp_dp` and `dT_dT` from the AD system should be `1` and the chain rule still holds.

Keep in mind that this API has to be independent from the primary variables used.  It looks a little bit that you are trying to invert the design and use the AD system to compute the non-AD value, like in `T_from_p_h`. That would create an inconsistency and this API would be usable only with p, h as primary variables. We do not want that.

May be, we need to look at what you are trying to do and how to fit it into this class.

Reply:
What I added here is basically a copy and paste from the definition in propfunc but with `p1` and `p2` renamed `p` and `T`.  Are you just saying that it should be declared as
```
 DualReal e_from_p_T(const DualReal & p1, const DualReal & p2) const
```
etc?

Reply:
I have just had a look at what I did here and I think that it implements exactly what you describe, and what is implemented in `propfunc` for all of the other methods. The input DualReals `p` and `T` can be functions of whatever with all of the derivatives carried through, and they are multiplied by the derivatives `de_dp` and `de_dT`  computed in the `e_from_p_T(p, T, e, de_dp, de_dT)` method.

`e_from_p_T` and `T_from_p_h` aren't currently declared using `propfunc` as they had default implementations, so currently there is no AD version of these methods (I guess that they aren't declared using `propfunc(e,p,T)` as you can't declare the methods twice?)

I do compute the (Real) derivatives in `WaterFluidProperties::T_from_p_h(p, h, T, dT_dp, dT_dh)` using a seperate AD method to avoid coding all the derivatives for all of the subregions, but the public AD method `T_from_p_h(p, h)` is unchanged, so should work with `p` and `h` defined in terms of other variables as you described above.

Reply:
> I think that it implements exactly what you describe, and what is implemented in `propfunc` for all of the other methods.

So, why did not you use the `propfunc` macro? When we add the derivative version, then the macro should be used.

Reply:
Because these methods are implemented in `SinglePhaseFluidProperties` (with a default implementation), they can't be declared using `propfunc` because they would be defined twice in the same class. 

I could add a `SinglePhaseFluidPropertiesBase` that has all of the methods declared using `propfunc` and then override the ones with default implementations in `SinglePhaseFluidProperties` to get around this I guess.

Reply:
I finally understand why you did it this way.  I forgot that the default implementation of `e_from_p_T` is actually a useful one. It does not throw the error of not being implemented...

Sorry it took so long...

Reply:
No worries David